notion-core/src/main/scala/com/ambiata/notion/core/LocationIO.scala

```diff
@@ -110,6 +109,19 @@ case class LocationIO(configuration: Configuration, s3Client: AmazonS3Client) {
       }}.as(state)
     }
 
+  /** read a file as a stream of bytes and compute some state value */
+  def streamBytes[A](location: Location, empty: => A, bufferSize: Int = 4096)(f: (Buffer, A) => A): RIO[A] =
+    RIO.io(empty).flatMap { s =>
+      var state = s
+      readUnsafe(location) { in => RIO.io {
+        val buffer = Buffer.wrapArray(Array.ofDim[Byte](bufferSize), 0, bufferSize)
+        var length = 0
+        while ({ length = in.read(buffer.bytes, buffer.offset, bufferSize); length != -1 }) {
+          state = f(Buffer.allocate(buffer, length), state)
```

Comment:
> there could just be "setLength" on Buffer that is guaranteed to re-use the same buffer or fail.

I think I'm OK with with making `buffer` a `var` for now, just to avoid adding functions that people might be tempted to call that they should avoid. At least until we know how common/useful it might be.